File: app/src/main/java/com/example/whattowatch/ui/presenter/DetailPresenter.java
package com.example.whattowatch.ui.presenter;

import android.util.Log;

import androidx.room.Insert;

import com.arellomobile.mvp.InjectViewState;
import com.arellomobile.mvp.MvpPresenter;
import com.example.whattowatch.App;
import com.example.whattowatch.database.MovieDao;
import com.example.whattowatch.model.mymodel.MyCastModel;
import com.example.whattowatch.model.mymodel.MyDetailModel;
import com.example.whattowatch.model.mymodel.MyVideoModel;
import com.example.whattowatch.repository.MoviesRepo;
import com.example.whattowatch.ui.view.DetailMovieView;
import com.google.android.material.snackbar.Snackbar;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import io.reactivex.Completable;
import io.reactivex.Flowable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

@InjectViewState
public class DetailPresenter extends MvpPresenter<DetailMovieView> {

    private int MOVIE_ID;
    private MyDetailModel data = null;
    private List<MyVideoModel> videos = new ArrayList<>();
    private List<MyCastModel> casts = new ArrayList<>();

    @Inject
    public MoviesRepo moviesRepo;

    private boolean isSaved = false;
    private CompositeDisposable compositeDisposable = new CompositeDisposable();

    public DetailPresenter(int id){
        App.getInstance().getAppComponent().inject(this);
        MOVIE_ID = id;
    }

    public DetailPresenter(){}

    @Override
    protected void onFirstViewAttach() {
        super.onFirstViewAttach();
        checkSavedInDb();
        getViewState().showProgress(true);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        compositeDisposable.clear();
    }

    @Override
    public void attachView(DetailMovieView view) {
        super.attachView(view);
        if(data == null) {
            loadDetailMovie(MOVIE_ID);

        }
    }

    public void add2Favorites(){

        if(isSaved){
            moviesRepo.deleteFromFavorites(data)
                    .subscribeOn(Schedulers.io())
                    .subscribe();
            isSaved = false;
        }else {
            Disposable d = moviesRepo.add2Favorites(data)
                    .subscribeOn(Schedulers.io())
                    .subscribe();
            compositeDisposable.add(d);
            isSaved = true;
        }
        getViewState().onSave(isSaved);


    }

    public void loadDetailMovie(int id) {
        Disposable d = moviesRepo.getDetailMovieInfo(id)
                  .subscribeOn(Schedulers.io())
                  .observeOn(AndroidSchedulers.mainThread())
                  .subscribe(
                          detailInfo ->{
                              getViewState().showProgress(false);
                             data = detailInfo;
                             getViewState().showMovieInfo(detailInfo);
                          },
                          error ->{
                              getViewState().showProgress(false);
                              getViewState().showErrorInfo();
                          }
                  );
        compositeDisposable.add(d);
                    d = moviesRepo.getMovieTrailers(id)
                            .subscribeOn(Schedulers.io())
                            .observeOn(AndroidSchedulers.mainThread())
                            .subscribe(movieTrailer ->{
                                videos = movieTrailer;
                                getViewState().showMovieTrailer(movieTrailer);
                            }, error->{
                                getViewState().showErrorTrailer();
                            } );
        compositeDisposable.add(d);

                    d = moviesRepo.getMovieCasts(id)
                            .subscribeOn(Schedulers.io())
                            .observeOn(AndroidSchedulers.mainThread())
                            .subscribe(castModels -> {
                                    casts = castModels;
                                    getViewState().showMovieCast(casts);
                                    },error ->{
                                        getViewState().showErrorCast();
                                    }
                            );
                    compositeDisposable.add(d);
    }

   private void checkSavedInDb(){
       Disposable d = moviesRepo.checkSavedFavorites(MOVIE_ID)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe( itIsSaved ->{
                    isSaved = true;
                        }
                );
       compositeDisposable.add(d);
    }
}

File: app/src/main/java/com/example/whattowatch/ui/presenter/DetailPresenter.kt
package com.example.whattowatch.ui.presenter

import com.arellomobile.mvp.InjectViewState
import com.arellomobile.mvp.MvpPresenter
import com.example.whattowatch.App
import com.example.whattowatch.App.Companion.instance
import com.example.whattowatch.model.mymodel.MyCastModel
import com.example.whattowatch.model.mymodel.MyDetailModel
import com.example.whattowatch.model.mymodel.MyVideoModel
import com.example.whattowatch.repository.MoviesRepo
import com.example.whattowatch.ui.view.DetailMovieView
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import java.util.*
import javax.inject.Inject

@InjectViewState
class DetailPresenter : MvpPresenter<DetailMovieView?> {
    private var MOVIE_ID = 0
    private var data: MyDetailModel? = null
    private var videos: List<MyVideoModel> = ArrayList()
    private var casts: List<MyCastModel> = ArrayList()
    @Inject
    lateinit var moviesRepo: MoviesRepo
    private var isSaved = false
    private val compositeDisposable = CompositeDisposable()

    constructor(id: Int) {
        MOVIE_ID = id
    }
    init {
        App.appComponent?.inject(this)
    }

    constructor() {}

    override fun onFirstViewAttach() {
        super.onFirstViewAttach()
        checkSavedInDb()
        viewState!!.showProgress(true)
    }

    override fun onDestroy() {
        super.onDestroy()
        compositeDisposable.clear()
    }

    override fun attachView(view: DetailMovieView?) {
        super.attachView(view)
        if (data == null) {
            loadDetailMovie(MOVIE_ID)
        }
    }

    fun add2Favorites() {
        isSaved = if (isSaved) {
            moviesRepo.deleteFromFavorites(data)
                    .subscribeOn(Schedulers.io())
                    .subscribe()
            false
        } else {
            val d = data?.let {
                moviesRepo!!.add2Favorites(it)
                        .subscribeOn(Schedulers.io())
                        .subscribe()
            }
            if (d != null) {
                compositeDisposable.add(d)
            }
            true
        }
        viewState!!.onSave(isSaved)
    }

    fun loadDetailMovie(id: Int) {
        var d = moviesRepo!!.getDetailMovieInfo(id)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        { detailInfo: MyDetailModel? ->
                            viewState!!.showProgress(false)
                            data = detailInfo
                            viewState!!.showMovieInfo(detailInfo)
                        }
                ) { error: Throwable? ->
                    viewState!!.showProgress(false)
                    viewState!!.showErrorInfo()
                }
        compositeDisposable.add(d!!)
        d = moviesRepo!!.getMovieTrailers(id)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ movieTrailer: List<MyVideoModel> ->
                    videos = movieTrailer
                    viewState!!.showMovieTrailer(movieTrailer)
                }) { error: Throwable? -> viewState!!.showErrorTrailer() }
        compositeDisposable.add(d)
        d = moviesRepo!!.getMovieCasts(id)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ castModels: List<MyCastModel> ->
                    casts = castModels
                    viewState!!.showMovieCast(casts)
                }
                ) { error: Throwable? -> viewState!!.showErrorCast() }
        compositeDisposable.add(d)
    }

    private fun checkSavedInDb() {
        val d = moviesRepo!!.checkSavedFavorites(MOVIE_ID)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe { itIsSaved: MyDetailModel? -> isSaved = true }
        compositeDisposable.add(d)
    }
}